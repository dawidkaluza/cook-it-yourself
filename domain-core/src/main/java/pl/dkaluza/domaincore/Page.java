package pl.dkaluza.domaincore;

import java.util.List;

import static pl.dkaluza.domaincore.Validator.validator;

public class Page<T> {
    private final List<T> items;
    private final int pageNumber;
    private final int totalPages;

    private Page(List<T> items, int pageNumber, int totalPages) {
        this.items = items;
        this.pageNumber = pageNumber;
        this.totalPages = totalPages;
    }

    public static <T> Factory<Page<T>> of(List<T> items) {
        return new PageFactory<>(items, 1, 1);
    }

    public static <T> Factory<Page<T>> of(List<T> items, int pageNumber, int totalPages) {
        return new PageFactory<>(items, pageNumber, totalPages);
    }

    public List<T> getItems() {
        return items;
    }

    public int getPageNumber() {
        return pageNumber;
    }

    public int getTotalPages() {
        return totalPages;
    }

    private static class PageFactory<T> extends DefaultFactory<Page<T>> {
        private PageFactory(List<T> items, int pageNumber, int totalPages) {
            super(
                ValidationExecutor.builder()
                    .withValidator(validator(items != null, "items", "Items must not be null"))
                    .withValidator(validator(pageNumber > 0, "pageNumber", "Page number must be a positive number"))
                    .withValidator(validator(totalPages > 0, "totalPages", "Total pages must be a positive number"))
                    .build(),
                () -> new Page<>(items, pageNumber, totalPages)
            );
        }
    }
}
