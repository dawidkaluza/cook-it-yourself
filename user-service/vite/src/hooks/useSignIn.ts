import {useCallback, useState} from "react";
import {userService} from "../domain/userService.ts";
import {useNavigate} from "react-router-dom";
import {InvalidCredentialsError} from "../domain/errors/user.tsx";
import {SignInRequest} from "../domain/dtos/user.ts";

export const useSignIn = () => {
  const [loading, setLoading] = useState(false);
  const [success, setSuccess] = useState("");
  const [error, setError] = useState("");
  const navigate = useNavigate();

  const signIn = useCallback(
    (signInReq: SignInRequest) => {
      setLoading(true);
      setSuccess("");
      setError("");

      userService.signIn(signInReq)
        .then(response => {
          setSuccess("Signing in proceeded successfully.");

          if (response.external) {
            setTimeout(() => {
              window.location.assign(response.redirectUrl);
            }, 800);
          } else {
            setTimeout(() => {
              navigate(response.redirectUrl);
            }, 800);
          }
        })
        .catch(error => {
          if (error instanceof InvalidCredentialsError) {
            setError("Invalid email or password.");
          } else {
            setError("Unable to process the request. Try again later.");
            console.error("Unable to process the request.", error);
          }
        })
        .finally(() => {
          setLoading(false)
        });

    }, [setLoading, setSuccess, setError, userService, setTimeout, window.location.assign, navigate]
  )

  return {
    signIn,
    loading,
    error,
    success,
  }
};