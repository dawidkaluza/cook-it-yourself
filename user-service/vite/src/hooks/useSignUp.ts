import {useCallback, useState} from "react";
import {useNavigate} from "react-router-dom";
import {userService} from "../domain/userService.ts";
import {SignUpRequest} from "../domain/dtos/user.ts";
import {InvalidFieldsError} from "../domain/errors/user.tsx";

export const  useSignUp = () => {
  const [loading, setLoading] = useState(false);
  const [success, setSuccess] = useState("");
  const [error, setError] = useState("");
  const [fieldsErrors, setFieldsErrors] = useState<Record<string, string>>({});
  const navigate = useNavigate();

  const signUp = useCallback((request: SignUpRequest) => {
    setLoading(true);
    setSuccess("");
    setError("");
    setFieldsErrors({});

    userService.signUp(request)
      .then(() => {
        setSuccess("Signed up successfully. You will be redirected to sign in page...");
        setTimeout(() => navigate("/sign-in"), 800);
      })
      .catch(error => {
        if (error instanceof InvalidFieldsError) {
          setFieldsErrors(error.fields);
        } else {
          setError("Unable to process the request. Try again later.");
          console.error("Unable to process the request.", error);
        }
      })
      .finally(() => {
        setLoading(false);
      });
  }, [setSuccess, setError, setLoading, setFieldsErrors, navigate]);

  return {
    signUp,
    loading,
    error,
    fieldsErrors,
    success
  }
};