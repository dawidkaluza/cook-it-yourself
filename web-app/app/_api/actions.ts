"use server"

import {cookies} from "next/headers";

export function isSignedIn(){
  const cookieStore = cookies();
  const sessionCookie = process.env.NEXT_PUBLIC_API_GATEWAY_SESSION_COOKIE;
  return sessionCookie ? cookieStore.has(sessionCookie) : false;
}