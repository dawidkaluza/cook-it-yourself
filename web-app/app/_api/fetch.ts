import {redirect} from "next/navigation";
import {cookies} from "next/headers";

type ApiRequest = {
  endpoint: string;
  method?: string;
  body?: any;
  headers?: object;
  ignoreAuth?: boolean;
};

type InternalApiRequest = ApiRequest & { baseUrl: string };

export class ApiError extends Error {
  private _response : Response
  constructor(response : Response) {
    super("Client error " + response.status);
    this._response = response;
  }

  get response(): Response {
    return this._response;
  }

  set response(value: Response) {
    this._response = value;
  }
}

function fetchApi({ baseUrl, endpoint, method, body, headers, ignoreAuth } : InternalApiRequest): Promise<any> {
  return new Promise((resolve, reject) => {
    fetch(
      baseUrl + endpoint,
      {
        method: method ?? "GET",
        body: body ? JSON.stringify(body) : undefined,
        headers: {
          "Accept": "application/json",
          "Content-Type": "application/json",
          ...headers,
        },
      }
    ).then(response => {
      if (response.ok) {
        if (!response.body) {
          resolve(null);
        } else {
          response.json()
            .then(data => {
              resolve(data)
            });
        }
      } else {
        switch (response.status) {
          case 401: {
            if (ignoreAuth) {
              reject(new ApiError(response));
            } else {
              redirect("/sign-in");
            }
            break;
          }

          case 403: {
            if (ignoreAuth) {
              reject(new ApiError(response));
            } else {
              redirect("/sign-in");
            }
            break;
          }

          default: {
            reject(new ApiError(response) );
            break;
          }
        }
      }
    }).catch(error => {
      reject(error);
    });
  });
}

export async function fetchFromServer(request : ApiRequest) {
  const { headers } = request;
  const baseUrl = process.env.API_GATEWAY_SERVER_URL;
  if (!baseUrl) {
    throw new Error("API_GATEWAY_SERVER_URL env var not defined");
  }

  return await fetchApi({
    ...request,
    baseUrl,
    headers: {
      ...headers,
      "Cookie": cookies().toString()
    }
  });
}

export async function fetchFromClient(request : ApiRequest) {
  const baseUrl = process.env.NEXT_PUBLIC_API_GATEWAY_CLIENT_URL;
  if (!baseUrl) {
    throw new Error("NEXT_PUBLIC_API_GATEWAY_CLIENT_URL env var not defined");
  }

  return await fetchApi({
    ...request,
    baseUrl,
  });
}
