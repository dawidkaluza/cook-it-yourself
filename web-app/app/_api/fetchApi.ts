function fetchApi(endpoint: string, method?: string, body?: any, headers?: object): Promise<any> {
  return new Promise((resolve, reject) => {
    let baseUrl;

    const isClient = typeof window !== 'undefined';
    if (isClient) {
      baseUrl = process.env.NEXT_PUBLIC_API_GATEWAY_CLIENT_URL;
    } else {
      baseUrl = process.env.API_GATEWAY_SERVER_URL;
    }

    fetch(
      baseUrl + endpoint,
      {
        method: method ?? "GET",
        body: body ? JSON.stringify(body) : undefined,
        headers: {
          "Accept": "application/json",
          "Content-Type": "application/json",
          ...headers,
        },
      }
    ).then(response => {
      if (response.ok) {
        if (!response.body) {
          resolve(null);
        } else {
          response.json()
            .then(data => resolve(JSON.parse(data)));
        }
      } else {
        switch (response.status) {
          case 401: {
            console.error("Unauthorized");
            reject({
              unauthorized: true,
              response,
            });
            break;
          }

          case 403: {
            console.error("Access denied");
            reject({
              accessDenied: true,
              response,
            });
            break;
          }

          default: {
            reject({ response });
            break;
          }
        }
      }
    }).catch(error => {
      reject(error);
    });
  });
}

export { fetchApi };