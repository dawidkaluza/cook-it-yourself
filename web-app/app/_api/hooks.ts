import {useCallback, useEffect, useState} from "react";

export const useAuth = (isAlreadySignedIn?: boolean) => {
  const [isSignedIn, setSignedIn] = useState(!!isAlreadySignedIn);
  const [name, setName] = useState<string | undefined>(undefined);

  useEffect(() => {
    if (!isAlreadySignedIn) {
      const name = sessionStorage.getItem("name");
      setSignedIn(name !== null);
      setName(name ?? undefined);
    }
  }, []);

  const signIn = useCallback((name?: string) => {
    setSignedIn(true);
    setName(name);
    sessionStorage.setItem("name", name ?? "");
  }, []);

  const signOut = useCallback(() => {
    sessionStorage.removeItem("name");
    setSignedIn(false);
    setName(undefined);
  }, []);

  return { isSignedIn, name, signIn, signOut };
};