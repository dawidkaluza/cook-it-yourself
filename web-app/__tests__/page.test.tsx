import {render, screen} from "@testing-library/react";
import Page from "@/app/page";
import {useAuth} from "@/app/_api/hooks";
import {Mock} from "vitest";


vi.mock("@/app/_api/hooks", () => {
  return {
    useAuth: vi.fn(),
  };
});

describe("page component", () => {
  test("render when signed out", () => {
    // Given
    (useAuth as Mock).mockImplementation(() => {
      return {
        isSignedIn: false,
        name: undefined
      };
    });

    // When
    const page = render(<Page />);

    // Then
    const signInLink = screen.getByRole("link", { name: /sign in/i });
    expect(signInLink).not.toBeNull();
    expect(signInLink?.getAttribute("href")).toBe("/sign-in");

    expect(page.container).toMatchSnapshot();
  });

  test("render when signed in", () => {
    // Given
    (useAuth as Mock).mockImplementation(() => {
      return {
        isSignedIn: true,
        name: "Dawid"
      };
    });

    // When
    const page = render(<Page />);

    // Then
    const welcomeText = screen.getByText(/welcome/i);
    expect(welcomeText).not.toBeNull();

    expect(page.container).toMatchSnapshot();
  });
});