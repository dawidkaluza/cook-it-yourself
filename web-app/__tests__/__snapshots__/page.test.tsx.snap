// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`page component > render when signed in 1`] = `
<div>
  <p
    class="text-center"
  >
    Welcome, 
    Dawid
    !
  </p>
  <p
    class="text-center"
  >
    Use the navigation above to start your cooking journey.
  </p>
</div>
`;

exports[`page component > render when signed out 1`] = `
<div>
  <p
    class="text-center"
  >
    Manage recipes, rate dishes, plan diet, plan grocery shopping and more.
  </p>
  <p
    class="text-center"
  >
    Want to try it out? 
    <a
      href="/sign-in"
    >
      Sign in
    </a>
    .
  </p>
</div>
`;
